### ISSUE TITLE
Formatted content causes unexpected keyboard navigation behavior.

### ISSUE THREAD

**Issue by mor10** [WordPress/gutenberg#4300]
## Issue Overview
When the first word of a block has formatting (eg **bold** or _italicized_) applied, keyboard navigation using arrow keys is broken.

## Steps to Reproduce (for bugs)

1. Create two paragraph blocks with some content. Can just be a single line each.
2. Highlight the first word or full line in the second block and italicize the text.
3. Use left arrow keys to navigate out of the block with the first word italicized and up to the previous one.
4. Use right arrow key to navigate back to the block with italicized text.

## Expected Behavior

Using left/right arrow keys to navigate between text in blocks should be seamless, and is when first word of the next block has no inline formatting.

## Current Behavior

If the first word in a block is formatted and the user navigates out of the block using the left arrow key, using the right arrow key to navigate back in doesn't work.

## Possible Solution

There is something odd happening in the interaction between the edge of a block and the edge of inline formatted content at the beginning of that block.

## Screenshots / Video

https://cloudup.com/cYiiKSh5Owa

**Comment by mcsf** [comment]
Thanks for the thorough report. It's a bit late for me here, but this is good thing for me or @iseulde to have a look at soon.

**Comment by iseulde** [comment]
I believe this is covered by #3015? Cc @ephox-mogran. I'll have another look at that PR.

**Comment by iseulde** [comment]
Oh I see you already linked the issue version, @mcsf. I think this is a duplicate?

**Comment by iseulde** [comment]
Hm, I think it might be worth splitting that PR up a bit. There's some duplicate logic in the writing flow Dom utils and editable: `isStartOfEditor` and `isEndOfEditor`. Additionally both don't account for the formatting/link boundaries. There **is** some logic for the boundaries in `Editable` though, but not for edge detection and it is specific to links: `isLinkBoundary`. This should ideally all be combined. :)

**Comment by iseulde** [comment]
@mor10 Which browser are you using? What I _can_ reproduce is the caret moving to the other block when you're still inside a formatting boundary. In other words, the caret should first move outside the boundary before moving on the the next block. I cannot reproduce the caret not going back to the previous block, so I'm curious what browser you are using.

**Comment by mor10** [comment]
Chrome, on Mac. I've tested it across two different computers. Same behavior.

**Comment by iseulde** [comment]
Thanks @mor10. Couldn't reproduce with the demo content, but did with a new post. #4358 should fix this issue.
